pkg/apis/serving/v1alpha1/revision_validation.go

```diff
@@ -118,24 +131,78 @@ func ValidateContainerConcurrency(cc RevisionContainerConcurrencyType, cm Revisi
 	return nil
 }
 
-func validateContainer(container corev1.Container) *apis.FieldError {
+func validateVolume(volume corev1.Volume) *apis.FieldError {
+	var errs *apis.FieldError
+	if volume.Name == "" {
+		errs = apis.ErrMissingField("name")
+	} else if len(validation.IsDNS1123Label(volume.Name)) != 0 {
+		errs = apis.ErrInvalidValue(volume.Name, "name")
+	}
+
+	vs := volume.VolumeSource
+	switch {
+	case vs.Secret != nil, vs.ConfigMap != nil:
+		// These are fine.
+	default:
+		errs = errs.Also(apis.ErrMissingOneOf("secret", "configMap"))
+	}
+	return errs
+}
+
+func validateContainer(container corev1.Container, volumes sets.String) *apis.FieldError {
 	if equality.Semantic.DeepEqual(container, corev1.Container{}) {
 		return apis.ErrMissingField(apis.CurrentField)
 	}
+
+	// Check that volume mounts match names in "volumes", that "volumes" has 100%
+	// coverage, and the field restrictions.
+	seen := sets.NewString()
+	var errs *apis.FieldError
+	for i, vm := range container.VolumeMounts {
+		// This effectively checks that Name is non-empty because Volume name must be non-empty.
+		if !volumes.Has(vm.Name) {
+			errs = errs.Also((&apis.FieldError{
+				Message: "volumeMount has no matching volume",
+				Paths:   []string{"name"},
+			}).ViaFieldIndex("volumeMounts", i))
+		}
+		seen.Insert(vm.Name)
+
+		if vm.MountPath == "" {
+			errs = errs.Also(apis.ErrMissingField("mountPath").ViaFieldIndex("volumeMounts", i))
+		}
+		if !vm.ReadOnly {
+			errs = errs.Also(apis.ErrMissingField("readOnly").ViaFieldIndex("volumeMounts", i))
+		}
+
+		if vm.SubPath != "" {
```

Comment:
The documentation in `spec.md` indicates that unspecified fields (including those here) are disallowed.

```yaml
  # This specifies where the volumes present in the RevisionSpec will be mounted into                                                                                                                                        
  # the container.  Each of the volumes in the RevisionSpec must be mounted here                                                                                                                                             
  # or it is an error.                                                                                                                                                                                                       
  volumeMounts:                                                                                                                                                                                                              
  - name: ... # This must match a name from Volumes                                                                                                                                                                          
    mountPath: ... # Where to mount the named Volume.                                                                                                                                                                        
    readOnly: ... # Must be True, will default to True, so it may be omitted.                                                                                                                                                
    # All other fields are disallowed.
```